manpages/man/tiledb_array_export.1
.TH tiledb_array_export 1 "12 October 2015" "Version 0.1" "TileDB programs"
 
.SH NAME
tiledb_array_export - exports the data of an array into a CSV or binary file

.SH SYNOPSIS
.B tiledb_array_export 
.BI "-A " "array_name " "-f " "filename" 
.br
[
.BI "-w " "workspace "
] [
.BI "-g " "group "
] 
.br
[
.BI "-a " "attribute_name_1,..."
] [
.BI "-d " "dim_name_1,..."
] [
.BI "-r " "range"
] [ 
.BI "-F " format"
] [
.BI "-l " delimiter"
] [
.BI "-p " precision"
]

.SH DESCRIPTION
.B tiledb_export_csv
exports the data of an array into a CSV or binary file. The documentation on
the exported CSV or binary format can be found in program 
\fBtiledb_array_load\fR(1).  

The user can specify a subset of dimensions and attributes to export, as well as
to constrain the exported cell to a subarray. In addition, the user may specify 
whether to export the cells in the normal cell order they are stored or in 
\fBreverse\fR order. The user may also select to export the array in \fBdense\fR
format, in which case even the empty cells are going to be exported, having 
special "empty" values in their attributes (e.g., 0 for numerics, and 
NULL_CHAR, '*', for strings). Finally, the user may specify a custom delimiter, 
as well as the number of decimal digits to be printed for real numbers, for
the case of exporting to CSV files.

On error, it returns \fB-1\fR. Moreover, if it is compiled in \fBVERBOSE\fR 
mode, it prints comprehensive error messages on \fIstderr\fR. On success, it 
returns \fB0\fR. 

The following conventions apply to the \fBSYNOPSIS\fR section:

.TS
tab (@);
c lx .
\fBbold text\fR @ type exactly as shown
\fIitalic text\fR @ replace with appropriate argument
[\fB\-a \fIarg\fR]@ any or all options within [ ] are optional
.TE

.SH OPTIONS
.TP
.BI "-w" " workspace" "\fR, " \fB --workspace=\fIworkspace\fR  
The path to the workspace folder where the array is stored. If it is not 
provided, the current working directory is set as the workspace by default.

.TP
.BI "-g" " group" "\fR, " \fB --group=\fIgroup\fR  
This is a directory inside the workspace where the array is stored.
Note that any group path provided is converted into an absolute path 
\fBrelative to the workspace directory\fR, i.e., regarding all home ("\\~/"), 
current ("./") and root ("/") as the workspace directory. If the group is not 
provided, the workspace is set as the default group.

.TP
.BI "-A" " array_name" "\fR, " \fB --array-name=\fIarray_name\fR  
The name of the array whose data will be exported. The array
must be already defined. 

.TP
.BI "-f" " filename" "\fR, " \fB --filename=\fIfilename\fR  
The name of the exported file. 

.TP
.BI "-a" " attribute_name_1,..." "\fR, " \
\fB --attribute-names=\fIattribute_name_1,...\fR  
The attributes to be exported. Any order and multiplicities are allowed. To
export \fBno\fR attribute, include \fB__hide\fR instead of a name list. To
export \fBall\fR attributes, omit the option.

.TP
.BI "-d" " dim_name_1,..." "\fR, " \
\fB --dim-names=\fIdim_name_1,...\fR  
The coordinates to be exported. Any order and multiplicities are allowed. To
export \fBno\fR coordinates, include \fB__hide\fR instead of a name list. To
export \fBall\fR coordinates, omit the option.

.TP
.BI "-r" " range" "\fR, " \fB --range=\fIrange\fR  
A range given as a sequence of [low,high] bounds across each dimension.
Each range bound must be a real number. The range constrains the exported
cells into a subarray.

.TP
.BI "-F" " format" "\fR, " \fB --format=\fIformat\fR  
It can be one of the following: 
\fBcsv\fR (CSV format), 
\fBcsv.gz\fR (GZIP-compressed CSV format), 
\fBdense.csv\fR (dense CSV format), 
\fBdense.csv.gz\fR (GZIP-compressed dense CSV format), 
\fBreverse.csv\fR (CSV format in reverse order), 
\fBreverse.csv.gz\fR (GZIP-compressed CSV format in reverse order), 
\fBreverse.dense.csv\fR (dense CSV format in reverse order), 
\fBreverse.dense.csv.gz\fR (dense GZIP-compressed CSV format in reverse order), 
\fBbin\fR (binary format), 
\fBbin.gz\fR (GZIP-compressed binary format), 
\fBdense.bin\fR (dense binary format), 
\fBdense.bin.gz\fR (GZIP-compressed dense binary format), 
\fBreverse.bin\fR (binary format in reverse order), 
\fBreverse.bin.gz\fR (GZIP-compressed binary format in reverse order), 
\fBreverse.dense.bin\fR (dense binary format in reverse order), 
\fBreverse.dense.bin.gz\fR (dense GZIP-compressed binary format in reverse 
order).
If it is not provided, then the program resolves the format by checking if
the file name contains any of the above suffices. For instance, "my_file.bin"
implies that the file is binary, whereas "my_file.csv.gz" implies that the file 
is CSV and compressed with GZIP. 

.TP
.BI "-l" " delimiter" "\fR, " \fB --delimiter=\fIdelimiter\fR  
This is meaningful only for CSV format. It stands for the delimiter which 
separates the values in a CSV line in the CSV file. If not given, the default is
\fB','\fR. For using a tab as a delimiter, the user must specify it simply 
as \fBtab\fR.

.TP
.BI "-p" " precision" "\fR, " \fB --precision=\fIprecision\fR  
This is meaningful only for CSV format. It should be an integer indicating the
number of decimal digits to be printed for real values. If not provided, the
default is \fB6\fR. 

.SH EXAMPLES
Assuming that the array has dimensions \fIdim1\fR,\fIdim2\fR and attributes 
\fIattr1\fR,\fIattr2\fR,\fIattr3\fR
.TP
\fBtiledb_array_export\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-g \fRmy_group/ \\
    \fB-A \fRmy_array \\
    \fB-f \fRmy_file.csv \\
    \fB-l \fR'|' \\
    \fB-p \fR3

Exports into \fImy_array.csv\fR all the cells of array \fImy_array\fR, stored in
folder \fImy_workspace/my_group/my_array\fR, in their native order, including 
all the coordinates and attributes. Also they are exported in sparse format.
The delimiter used for the values in each CSV line is '|'. The precision of
real values is limited to 3 decimal digits.

.TP
\fBtiledb_array_export\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-g \fRmy_group/ \\
    \fB-A \fRmy_array \\
    \fB-f \fRmy_file \\
    \fB-d \fRdim1 \\
    \fB-F \fRdense.bin.gz \\
    \fB-a \fRattr1,attr2

Exports into GZIP compressed binary file \fImy_file\fR all the cells of 
\fImy_array\fR, stored in folder \fImy_workspace/my_group/my_array\fR, in the 
native order, including only the coordinates on the \fIdim1\fR dimension and the
values on the \fIattr1\fR,\fIattr2\fR attributes. The array is 
exported in dense form, i.e., even the empty cell coordinates will appear in
the file, having "empty" attribute values.

.TP
\fBtiledb_array_export\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-g \fRmy_group/ \\
    \fB-A \fRmy_array \\
    \fB-f \fRmy_file \\
    \fB-d \fRdim1 \\
    \fB-a \fRattr1,attr2 \\
    \fB-F \fRreverse.dense.bin.gz

Same as above, but the cells are exported in the reverse order.

\fBtiledb_array_export\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-g \fRmy_group/ \\
    \fB-A \fRmy_array \\
    \fB-f \fRmy_file.dense.bin.gz \\
    \fB-d \fR__hide \\
    \fB-r \fR0,10,10,20 \\
    \fB-a \fRattr1,attr2 

Same as the second example, but no coordinate is exported, and only the cells
with dim1 in [0,10] and dim2 in [10,20] are exported. Also notice that the
file format is derived from the suffix of the file name.

.TP
\fBtiledb_array_export\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-g \fRmy_group/ \\
    \fB-A \fRmy_array \\
    \fB-f \fRmy_file \\
    \fB-d \fRdim1 \\
    \fB-F \fRdense.bin.gz \\
    \fB-a \fR__hide

Same as the second example, but no attribute value is exported.

.TP
\fBtiledb_array_export\fR \\ 
    \fB-w \fRmy_workspace/ \\
    \fB-g \fRmy_group/ \\
    \fB-A \fRmy_array \\
    \fB-f \fRmy_file \\
    \fB-d \fRdim1 \\
    \fB-F \fRdense.bin.gz \\
    \fB-a \fRattr1,attr2,attr1

Same as the second example, but now the \fIattr1\fR values are shown twice
(once before those of \fIattr2\fR and once after).

.SH EXIT STATUS
.TP 
\fB0\fR for success and \fB-1\fR for error.

.SH SEE ALSO
.BR "tiledb_array_load" "(1) " "tiledb_dataset_generate" "(1) "

.SH AUTHOR
Stavros Papadopoulos <http://people.csail.mit.edu/stavrosp>, Copyright (c) 2015


tiledb_cmd/src/tiledb_array_export.cc
/**
 * @file   tiledb_array_export_csv.cc
 * @author Stavros Papadopoulos <stavrosp@csail.mit.edu>
 *
 * @section LICENSE
 *
 * The MIT License
 * 
 * @copyright Copyright (c) 2015 Stavros Papadopoulos <stavrosp@csail.mit.edu>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * 
 * @section DESCRIPTION
 *
 * Implements command "tiledb_array_export".
 */

#include "csv_line.h"
#include "special_values.h"
#include "tiledb.h"
#include "utils.h"
#include <getopt.h>
#include <iostream>
#include <string>

#define PRINT_ERROR(x) std::cerr << "[TileDB] Error: " << x << ".\n"
#define PRINT(x) std::cout << "[TileDB] " << x << ".\n"

/** 
 * Parses the command options. It returns 0 on success. On error, it prints
 * a message on stderr and returns -1.
 */                                                            
int parse_options(
    int argc, 
    char** argv, 
    std::string& workspace,
    std::string& group,
    std::string& array_name, 
    std::string& filename,
    std::string& format,
    std::vector<std::string>& dim_names,
    std::vector<std::string>& attribute_names,
    double*& range,
    int& range_size,
    char& delimiter,
    int& precision) {

  // Initialization
  workspace = "";
  group = "";
  array_name = "";
  filename = "";
  std::string attribute_names_str("");
  std::string dim_names_str("");
  std::string range_str = "";
  std::string mode("");
  dim_names.clear();
  attribute_names.clear();
  format = "";
  std::string delimiter_str = "";
  std::string precision_str = "";

  // Auxiliary variable
  CSVLine temp;

  // *************** //
  // Parse options * //
  // *************** // 

  // Define long options
  static struct option long_options[] = 
  {
    {"attribute-names",1,0,'a'},
    {"array-name",1,0,'A'},
    {"dim-names",1,0,'d'},
    {"filename",1,0,'f'},
    {"format",1,0,'F'},
    {"group",1,0,'g'},
    {"delimiter",1,0,'l'},
    {"precision",1,0,'p'},
    {"range",1,0,'r'},
    {"workspace",1,0,'w'},
    {0,0,0,0},
  };

  // Define short options
  const char* short_options = "a:A:d:f:F:g:l:p:r:w:";
  // Get options
  int c;
  int option_num = 0;
  while((c=getopt_long(argc, argv, short_options, long_options, NULL)) >= 0) {
    ++option_num;
    switch(c) {
      case 'a':
        if(attribute_names_str != "") {
          PRINT_ERROR("More than one attribute name lists provided");
          return -1;
        }
        attribute_names_str = optarg;
        break;
      case 'A':
        if(array_name != "") {
          PRINT_ERROR("More than one array names provided");
          return -1;
        }
        array_name = optarg;
        break;
      case 'd':
        if(dim_names_str != "") {
          PRINT_ERROR("More than one dimension name lists provided");
          return -1;
        }
        dim_names_str = optarg;
        break;
      case 'f':
        if(filename != "") {
          PRINT_ERROR("More than one CSV file names provided");
          return -1;
        }
        filename = optarg;
        break;
      case 'F':
        if(format != "") {
          PRINT_ERROR("More than one formats provided");
          return -1;
        }
        format = optarg;
        break;
      case 'g':
        if(group != "") {
          PRINT_ERROR("More than one groups provided");
          return -1;
        }
        group = optarg;
        break;
      case 'l':
        if(delimiter_str != "") {
          PRINT_ERROR("More than one delimiters provided");
          return -1;
        }
        delimiter_str = optarg;
        break;
      case 'p':
        if(precision_str != "") {
          PRINT_ERROR("More than one precision values provided");
          return -1;
        }
        precision_str = optarg;
        break;
      case 'r':
        if(range_str != "") {
          PRINT_ERROR("More than one ranges provided");
          return -1;
        }
        range_str = optarg;
        break;
      case 'w':
        if(workspace != "") {
          PRINT_ERROR("More than one workspaces provided");
          return -1;
        }
        workspace = optarg;
        break; 
      default:
        return -1;
    }
  }

  // ****************** //
  // Check correctness  //
  // ****************** //

  // Check number of arguments
  if((argc-1) != 2*option_num) {
    PRINT_ERROR("Arguments-options mismatch");
    return -1;
  }
  //----- array name ----- //
  if(array_name == "") {
    PRINT_ERROR("Array name not provided");
    return -1;
  }
  temp << array_name;
  if(temp.val_num() > 1) {
    PRINT_ERROR("More than one array names provided");
    return -1;
  }
  // ----- workspace ----- //
  temp.clear();
  temp << workspace;
  if(temp.val_num() > 1) {
    PRINT_ERROR("More than one workspaces provided");
    return -1;
  }
  // ----- group ----- //
  temp.clear();
  temp << group;
  if(temp.val_num() > 1) {
    PRINT_ERROR("More than one groups provided");
    return -1;
  }
  // ----- filename ----- //
  if(filename == "") {
    PRINT_ERROR("File name not provided");
    return -1;
  }
  temp.clear();
  temp << filename;
  if(temp.val_num() > 1) {
    PRINT_ERROR("More than one file names provided");
    return -1;
  }
  // ----- format ----- //
  if(format == "") { // Format not given, derive it from filename
    if(ends_with(filename, ".reverse.dense.csv")) {
      format = "reverse.dense.csv"; 
    } else if(ends_with(filename, ".reverse.dense.csv.gz")) {
      format = "reverse.dense.csv.gz"; 
    } else if(ends_with(filename, ".reverse.csv")) {
      format = "reverse.csv"; 
    } else if(ends_with(filename, ".reverse.csv.gz")) {
      format = "reverse.csv.gz"; 
    } else if(ends_with(filename, ".dense.csv")) {
      format = "dense.csv"; 
    } else if(ends_with(filename, ".dense.csv.gz")) {
      format = "dense.csv.gz"; 
    } else if(ends_with(filename, ".csv")) {
      format = "csv"; 
    } else if(ends_with(filename, ".csv.gz")) {
      format = "csv.gz"; 
    } else if(ends_with(filename, ".reverse.dense.bin")) {
      format = "reverse.dense.bin"; 
    } else if(ends_with(filename, ".reverse.dense.bin.gz")) {
      format = "reverse.dense.bin.gz"; 
    } else if(ends_with(filename, ".reverse.bin")) {
      format = "reverse.bin"; 
    } else if(ends_with(filename, ".reverse.bin.gz")) {
      format = "reverse.bin.gz"; 
    } else if(ends_with(filename, ".dense.bin")) {
      format = "dense.bin"; 
    } else if(ends_with(filename, ".dense.bin.gz")) {
      format = "dense.bin.gz"; 
    } else if(ends_with(filename, ".bin")) {
      format = "bin"; 
    } else if(ends_with(filename, ".bin.gz")) {
      format = "bin.gz";
    } else {
      PRINT_ERROR("Cannot derive file format");
      return -1;
    }
  } else { // Format is given, check correctness
    if(format != "csv" && format != "csv.gz" &&
       format != "dense.csv" && format != "dense.csv.gz" &&
       format != "reverse.csv" && format != "reverse.csv.gz" &&
       format != "reverse.dense.csv" && format != "reverse.dense.csv.gz" &&
       format != "bin" && format != "bin.gz" &&
       format != "dense.bin" && format != "dense.bin.gz" &&
       format != "reverse.bin" && format != "reverse.bin.gz" &&
       format != "reverse.dense.bin" && format != "reverse.dense.bin.gz") {
      PRINT_ERROR("Invalid file format");
      return -1;
    }
  }
  // ----- delimiter ----- //
  if(delimiter_str == "") {
    delimiter = ',';
  } else if(ends_with(format, "bin") || ends_with(format, "bin.gz")) {
    PRINT_ERROR("The delimiter is meaningless for binary format");
    return -1;
  } else if(delimiter_str == "tab") {
      delimiter = '\t';
  } else if(delimiter_str.size() == 1) {
      delimiter = delimiter_str[0];
  } else {
      PRINT_ERROR("Invalid delimiter");
      return -1;
  }
  // ----- dimension names ----- //
  if(dim_names_str != "")
    dim_names = CSVLine(&dim_names_str[0]).values_str_vec();
  // ----- attribute names ----- //
  if(attribute_names_str != "")
    attribute_names = CSVLine(&attribute_names_str[0]).values_str_vec();
  // ----- range ----- //
  if(range_str == "") {
    range_size = 0;
    range = NULL;
  } else {
    std::vector<std::string> range_str_vec;
    range_str_vec = CSVLine(&range_str[0]).values_str_vec();
    range_size = int(range_str_vec.size());
    range = new double[range_size];
    for(int i=0; i<range_size; ++i) {
      if(!is_real(range_str_vec[i].c_str())) {
        PRINT_ERROR("The range bounds must be real numbers");
        delete [] range;
        return -1;
      }
      sscanf(range_str_vec[i].c_str(), "%lf", &range[i]); 
    }
  }   
  // ----- precision ----- //
  if(precision_str == "") {
    precision = 6; // Default
  } else {
    temp.clear();
    temp << precision_str;
    if(temp.val_num() > 1) {
      PRINT_ERROR("More than one precision values");
      return -1;
    }
    if(!is_non_negative_integer(precision_str.c_str())) {
      PRINT_ERROR("The precision value must be a non-negative integer");
      return -1;
    } else {
      sscanf(precision_str.c_str(), "%d", &precision); 
    }
  }

  // Success
  return 0;
}

int main(int argc, char** argv) {
  // Parse command line
  std::string workspace;
  std::string group;
  std::string array_name;
  std::string filename;
  std::string format;
  std::vector<std::string> dim_names;
  std::vector<std::string> attribute_names;
  const char** dim_names_c_str;
  const char** attribute_names_c_str;
  double* range = NULL;
  int dim_names_num, attribute_names_num, range_size, precision;
  char delimiter;

  if(parse_options(
         argc, argv, workspace, group, array_name, filename,
         format, dim_names, attribute_names,
         range, range_size, delimiter, precision))
    return -1;

  // Initialize TileDB
  TileDB_CTX* tiledb_ctx;
  if(tiledb_ctx_init(tiledb_ctx))
    return -1;

  // Get vectors for dimension and attribute names
  dim_names_num = int(dim_names.size());
  if(dim_names_num == 0) {
    dim_names_c_str = NULL;
  } else {
    dim_names_c_str = new const char*[dim_names_num];
    for(int i=0; i<dim_names_num; ++i) 
      dim_names_c_str[i] = dim_names[i].c_str();
  }
  attribute_names_num = int(attribute_names.size());
  if(attribute_names_num == 0) {
    attribute_names_c_str = NULL;
  } else {
    attribute_names_c_str = new const char*[attribute_names_num];
    for(int i=0; i<attribute_names_num; ++i) 
      attribute_names_c_str[i] = attribute_names[i].c_str();
  }

  // Export to CSV
  if(tiledb_array_export(
         tiledb_ctx, workspace.c_str(), group.c_str(),
         array_name.c_str(), filename.c_str(), format.c_str(),
         dim_names_c_str, dim_names_num, 
         attribute_names_c_str, attribute_names_num,
         range, range_size, delimiter, precision)) { 
    // Error - Clean up
    tiledb_ctx_finalize(tiledb_ctx);
    if(dim_names_c_str != NULL) 
      delete [] dim_names_c_str;
    if(attribute_names_c_str != NULL) 
      delete [] attribute_names_c_str;
    if(range != NULL)
      delete [] range;
    return -1;
  }

  // Clean up  
  if(dim_names_c_str != NULL) 
    delete [] dim_names_c_str;
  if(attribute_names_c_str != NULL) 
    delete [] attribute_names_c_str;
  if(range != NULL)
    delete [] range;

  // Finalize TileDB
  if(tiledb_ctx_finalize(tiledb_ctx))
    return -1;

  // Success
  PRINT("Program executed successfully");
  
  return 0;
}
